### PR TITLE
Watch rewrite

### PR DESCRIPTION
**Pull request by eatkins**
- [x] I've read the [CONTRIBUTING](https://github.com/sbt/sbt/blob/develop/CONTRIBUTING.md) guidelines

### CODE REVIEW

**Comment by eatkins** [comment]
This is built on top of #4476. Once that's merged in, I'll rebase this PR. While this is a fat change on it's own, it's bloated further by the ~1000 line diff due to #4476.

**Comment by eatkins** [comment]
Fixes #3718

**Comment by eatkins** [comment]
A few highlights:
1) There is now an alternative command `watch` that works the same as `~` but is implemented as an InputTask. This allows plugin authors to add watch functionality into a custom task. It also may be more clear for new users than `~`.
2) You can terminate a watch and return to the shell by typing 's'. This is useful if you ran, say, `sbt ~compile` but realized that you wanted to do something else. Before you'd have had to restart sbt entirely. Now you just easily drop back into an sbt shell.
3) Watch settings can be scoped at task level:
```
sbt:swoval> set Global / watchLogLevel := Level.Debug
[info] Defining Global / watchLogLevel
[info] The new value will be used by no settings or tasks.
[info] Reapplying settings...
[info] Set current project to swoval (in build file:/Users/ethanatkins/work/swoval/files/)
sbt:swoval> watch filesJVM/compile
[info] [watch] [debug] meta-build - Received Update(Entry(/Users/ethanatkins/work/swoval/files/.gitignore, Right(TypedPath(/Users/ethanatkins/work/swoval/files/.gitignore))),Entry(/Users/ethanatkins/work/swoval/files/.gitignore, Right(TypedPath(/Users/ethanatkins/work/swoval/files/.gitignore))),Deadline(645137216846528 nanoseconds))
[info] Formatting 63 sources in file-tree-views using JavaFormatter.
[info] Formatting 6 sources in file-tree-views using ClangFormatter.
[success] Total time: 1 s, completed Jan 11, 2019 2:11:24 PM
[info] [watch] 1. Waiting for source changes in project filesJVM... (press enter to interrupt, 'r' to re-run the command or 's' to return to the shell)
[info] [watch] [debug] meta-build - Received event for path /Users/ethanatkins/work/swoval/files/.gitignore
[info] [watch] [debug] Received input events: CancelWatch. Taking CancelWatch
[success] Total time: 2 s, completed Jan 11, 2019 2:11:26 PM
sbt:swoval> set filesJVM / watchLogLevel := Level.Info
[info] Defining filesJVM / watchLogLevel
[info] The new value will be used by no settings or tasks.
[info] Reapplying settings...
[info] Set current project to swoval (in build file:/Users/ethanatkins/work/swoval/files/)
sbt:swoval> watch filesJVM/compile
[info] Formatting 63 sources in file-tree-views using JavaFormatter.
[info] Formatting 6 sources in file-tree-views using ClangFormatter.
[success] Total time: 0 s, completed Jan 11, 2019 2:11:48 PM
[info] [watch] 1. Waiting for source changes in project filesJVM... (press enter to interrupt, 'r' to re-run the command or 's' to return to the shell)
[success] Total time: 4 s, completed Jan 11, 2019 2:11:50 PM
sbt:swoval> watch filesJS/compile
[info] [watch] [debug] meta-build - Received Update(Entry(/Users/ethanatkins/work/swoval/files/.gitignore, Right(TypedPath(/Users/ethanatkins/work/swoval/files/.gitignore))),Entry(/Users/ethanatkins/work/swoval/files/.gitignore, Right(TypedPath(/Users/ethanatkins/work/swoval/files/.gitignore))),Deadline(645167057517128 nanoseconds))
[info] stderr: info TOOL Using Ninja generator, because ninja is available.
[info] info CMD BUILD
[info] info RUN cmake --build "/Users/ethanatkins/work/swoval/files/files/js/npm/build" --config Release
[info]
[info] stdout: ninja: no work to do.
[info]
[success] Total time: 2 s, completed Jan 11, 2019 2:11:56 PM
[info] [watch] 1. Waiting for source changes in project filesJS... (press enter to interrupt, 'r' to re-run the command or 's' to return to the shell)
[info] [watch] [debug] meta-build - Received event for path /Users/ethanatkins/work/swoval/files/.gitignore
[info] [watch] [debug] Received input events: CancelWatch. Taking CancelWatch
[success] Total time: 9 s, completed Jan 11, 2019 2:12:02 PM
sbt:swoval> set Global / watchLogLevel := Level.Info
[info] Defining Global / watchLogLevel
[info] The new value will be used by no settings or tasks.
[info] Reapplying settings...
[info] Set current project to swoval (in build file:/Users/ethanatkins/work/swoval/files/)
sbt:swoval> watch filesJS/compile
[success] Total time: 0 s, completed Jan 11, 2019 2:12:14 PM
[info] [watch] 1. Waiting for source changes in project filesJS... (press enter to interrupt, 'r' to re-run the command or 's' to return to the shell)
[success] Total time: 5 s, completed Jan 11, 2019 2:12:18 PM
```
4) @japgolly There is now a watchOnEnter setting that can run arbitrary code before any of the watch processing begins. This can be used to clear the console.

**Comment by eatkins** [comment]
@OlegYch this PR resolves #3718 by introducing a new data structure `Glob` that looks like
```scala
sealed trait Glob {
  def base: File
  def filter: FileFilter
  def depth: Int
}
```
I added two setting keys, `inputs` and `triggers` that each have return type Seq[Glob]. Suppose I was adding a protoc command, then I could do it with:
```scala
val protoc = TaskKey[Unit]("compile protobuf files")
protoc / inputs += baseDirectory.value / "src" / "main" / "proto" ** "*.proto"
def protocImpl(files: Seq[File]): Unit = ???
protoc := {
  val sourceFiles = (protoc / inputs).value.all
  protocImpl(sourceFiles)
}
```
With these definitions `~protoc` will automatically track all of the "*.proto" files described by the `protoc / inputs` definition.
Internally this is how compile now works after these changes.

If you wanted to make protoc trigger automatically if a non-dependent file changes, that can be done with e.g.:
```scala
protoc / triggers += baseDirectory.value / "recompile-protobufs.txt"
```
This way the inputs and triggers are handled separately so that the protocImpl doesn't get the "recompile-protobufs.txt" file, but that file will still trigger a new build.

**Comment by eatkins** [comment]
fixes #4445

**Comment by OlegYch** [comment]
@eatkins this looks great to me
is there a plan of migrating core tasks to use it?

**Comment by eatkins** [comment]
@OlegYch I'm not sure how many core tasks really need to be migrated. I've already made compile work with the new infrastructure. Thus any task that depends on compile transitively works with the new watch infrastructure. I'm actually not aware of many built in tasks that depend on a set of source files. Any tasks that do, however, should be fairly straightforward to migrate.

Not directly related to watch, but I also use globs to improve io performance. I tacked on 
https://github.com/eatkins/sbt/commit/8983e0983b9c1201cb04556565af0bba422fce72 to this PR to update the clean task to use globs. The time to evaluate clean in the sbt project dropped from 6s to < 1s on my machine.

**Comment by eatkins** [comment]
I dropped the WIP from this PR. It still depends on #4539 (and its dependent PRs), but I consider it ready to go modulo code review feedback.

**Comment by eatkins** [comment]
@eed3si9n I consider this functionally complete. There are still some api changes I am making to glob and io that I will handle in a separate PR, but this would be usable for an M3.

**Review by eed3si9n — COMMENTED** [review]
The commit message on e2c971460a088b2c9e44500ac6bd5ded71194476 seems a bit outdated.

> One complicating factor is that dynamic tasks do not track any of
their dynamic dependencies. To work around this, I added the
transitiveDependencies key.

File: main/src/main/scala/sbt/Keys.scala
Comment (eed3si9n) [inline main/src/main/scala/sbt/Keys.scala]: 1. When would you need this, instead of `foo / inputs`?
2. `watchTriggers` might be a better name since the trigger might come up in different context like `def trigger = allRequirements`

File: main/src/main/scala/sbt/internal/FileManagement.scala
Comment (eed3si9n) [inline main/src/main/scala/sbt/internal/FileManagement.scala]: Is this undoing Windows specific handling added in https://github.com/sbt/sbt/commit/e3625ca8b0781288f379eff6d6e6534998d64200, or it's no longer needed?

**Comment by eed3si9n** [comment]
The commit message on d328d389cc7fd39eb75e4ea819a4e7721be7fc2d looks unfinished.

File: main/src/main/scala/sbt/internal/FileManagement.scala
Comment (eatkins) [reply main/src/main/scala/sbt/internal/FileManagement.scala]: It's no longer needed.

**Comment by eatkins** [comment]
Drat. I just independently noticed that I didn't finish the commit message to https://github.com/sbt/sbt/commit/d328d389cc7fd39eb75e4ea819a4e7721be7fc2d. I'm fixing the scripted tests now and will fix that.

File: main/src/main/scala/sbt/Keys.scala
Comment (eatkins) [reply main/src/main/scala/sbt/Keys.scala]: Yeah. That's a good point. I also am planning to move these keys into the plugin we discussed.

**Comment by eatkins** [comment]
> The commit message on [e2c9714](https://github.com/sbt/sbt/commit/e2c971460a088b2c9e44500ac6bd5ded71194476) seems a bit outdated.
> 
> > One complicating factor is that dynamic tasks do not track any of
> > their dynamic dependencies. To work around this, I added the
> > transitiveDependencies key.

Yeah. I should change it to say that that issue is addressed in subsequent commits.

File: main/src/main/scala/sbt/internal/FileManagement.scala
Comment (eatkins) [reply main/src/main/scala/sbt/internal/FileManagement.scala]: Actually, that is mistaken. It is not needed using new code that isn't part of this PR so I need to restore the windows specific code until I open that PR. It works fine on my VM, but CI fails with the regular file repository. I'm 90% sure that CI will pass with some changes from https://github.com/sbt/io/pull/226.

**Comment by eed3si9n** [comment]
AppVeyor failed, but I think it's unrelated Windows flakiness, so I'm going to rerun the job.

File: .appveyor.yml
Comment (eed3si9n) [inline .appveyor.yml]: We should split this commit out to another PR.
Maybe we should turn it off by default in CI.
https://github.com/sbt/sbt/blob/ca5f978df495a5a975302cedba73986d2abaa262/main/src/main/scala/sbt/Defaults.scala#L332

**Comment by eatkins** [comment]
I think the appveyor flakiness is real. It's due to io contention between the cache and the various task threads listing files and computing the file stamps. Linux and osx generally don't lock files which is why we don't see the errors on travis. I'm working on fixing it.

File: .appveyor.yml
Comment (eatkins) [reply .appveyor.yml]: Ok. I added #4590.  I'll remove this commit from this PR.

**Comment by eatkins** [comment]
I think the appveyor issues should be resolved now. They passed on my local appveyor instance at least.

**Review by eed3si9n — DISMISSED** [review]
LGTM

**Comment by eatkins** [comment]
Thanks for the approval! I had to resolve a merge conflict from #4590 so I need a re-approval for this pending CI.

**Review by eed3si9n — APPROVED** [review]

**Pull request merged**
